### PR TITLE
replace onlyOnCPUAndCUDA with onlyNativeDeviceTypes

### PR DESCRIPTION
**Pull request by kshitij12345**
Reference #53849

Replace `onlyOnCPUandCUDA` with `onlyNativeDeviceTypes` which includes `cpu, cuda and meta`.

### CODE REVIEW

File: test/test_binary_ufuncs.py [hunk 1]
```diff
@@ -532,7 +532,8 @@ def test_out_resize_warning(self, device):
                 self.assertEqual(len(w), 1)
 
     # Verifies that the inplace dunders (like idiv) actually are in place
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # UserWarning not triggered
```
Comment (kshitij12345) [inline test/test_binary_ufuncs.py:535]: Note the expected failure

File: test/test_nn.py [hunk 2]
```diff
@@ -14323,7 +14323,8 @@ def test_adaptive_avg_pool3d_output_size_one(self, device):
         self.assertEqual(out.stride(), [c, 1, 1, 1, 1])
 
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # Runtime Error not raised for meta
```
Comment (kshitij12345) [inline test/test_nn.py:14326]: Note the expected failure

File: test/test_nn.py [hunk 3]
```diff
@@ -16698,7 +16699,8 @@ def test_grid_sample_nan_inf(self, device, dtype):
                                                      padding_mode=padding_mode, align_corners=False)
             self.assertEqual(sample, torch.zeros([1, 1, 1, 2], device=device, dtype=dtype))
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # RuntimeError: Unrecognized tensor type ID: Meta
```
Comment (kshitij12345) [inline test/test_nn.py:16702]: Note the expected failure

File: test/test_nn.py [hunk 4]
```diff
@@ -16735,7 +16737,8 @@ def func(x):
                 # Incorrect output_size
                 F.fractional_max_pool2d(x, (2, 2), output_size=output_size, _random_samples=samples)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # RuntimeError: Unrecognized tensor type ID: Meta
```
Comment (kshitij12345) [inline test/test_nn.py:16740]: Note the expected failure

File: test/test_ops.py [hunk 5]
```diff
@@ -54,8 +54,9 @@ def tearDownClass(cls):
 
     # Validates that each OpInfo specifies its forward and backward dtypes
     #   correctly for CPU and CUDA devices
+    @skipMeta
```
Comment (kshitij12345) [inline test/test_ops.py:57]: Note the skip (some tests pass and some tests fail). Hence the skip

File: test/test_reductions.py [hunk 6]
```diff
@@ -854,7 +854,8 @@ def testset_for_shape(shape, i):
         # Naive kernel for big slice sizes (> 2048)
         testset_for_shape((10, 4096), 10)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # mode only supports CPU and CUDA device type
```
Comment (kshitij12345) [inline test/test_reductions.py:857]: Note the expected failure

File: test/test_reductions.py [hunk 7]
```diff
@@ -1022,8 +1023,8 @@ def test_amin(self, device, dtype):
     def test_amax(self, device, dtype):
         self._test_minmax_helper(torch.amax, np.amax, device, dtype)
 
-    @onlyOnCPUAndCUDA
-    @dtypesIfCPU(torch.float, torch.double)
+    @onlyNativeDeviceTypes
+    @dtypes(torch.float, torch.double)
```
Comment (kshitij12345) [inline test/test_reductions.py:1027]: Note dtypesIfCPU -> dtypes

File: test/test_shape_ops.py [hunk 8]
```diff
@@ -224,9 +224,9 @@ def test_diagonal_multidim(self, device, dtype):
         self.assertEqual(expected.shape, result.shape)
         self.assertEqual(expected, result)
 
-    @onlyOnCPUAndCUDA
-    @dtypesIfCPU(*get_all_dtypes(include_complex=False, include_bool=False, include_half=False,
-                                 include_bfloat16=False))
+    @onlyNativeDeviceTypes
+    @dtypes(*get_all_dtypes(include_complex=False, include_bool=False, include_half=False,
```
Comment (kshitij12345) [inline test/test_shape_ops.py:228]: Note dtypesIfCPU -> dtypes

File: test/test_tensor_creation_ops.py [hunk 9]
```diff
@@ -2783,11 +2783,11 @@ def test_tensor_ctor_device_inference(self, device):
                                                             sparse_size, dtype=torch.float64)
                 self.assertEqual(sparse_with_dtype.device, torch.device('cpu'))
 
-    @onlyOnCPUAndCUDA
+    @onlyNativeDeviceTypes
     @precisionOverride({torch.bfloat16: 5e-2, torch.half: 1e-3})
     @unittest.skipIf(not TEST_SCIPY, "Scipy not found")
     @dtypesIfCUDA(torch.float, torch.double, torch.bfloat16, torch.half, torch.long)
-    @dtypesIfCPU(torch.float, torch.double, torch.long)
+    @dtypes(torch.float, torch.double, torch.long)
```
Comment (kshitij12345) [inline test/test_tensor_creation_ops.py:2790]: Note dtypesIfCPU -> dtypes

File: test/test_testing.py [hunk 10]
```diff
@@ -585,7 +585,8 @@ def test_trivial_passing_test_case_on_cpu_cuda(self, device):
         # we are currently disabling CUDA early termination for distributed tests.
         self.assertIn('Ran 2 test', stderr)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # This is only supported for CPU and CUDA
```
Comment (kshitij12345) [inline test/test_testing.py:588]: Note expectedFailureMeta

File: test/test_torch.py [hunk 11]
```diff
@@ -4026,7 +4027,8 @@ def forward_func(slf, device):
         test_func(torch.Tensor.scatter_add)
         test_func(torch.scatter_add)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # expected a non-determinitic error, but it was not raised
```
Comment (kshitij12345) [inline test/test_torch.py:4030]: Note expectedFailureMeta

File: test/test_torch.py [hunk 12]
```diff
@@ -6451,8 +6453,9 @@ def test_ternary_op_mem_overlap(self, device, dtype):
             self.ternary_check_input_output_mem_overlap(out_op, dev,
                                                         expected_failure=not has_input_output_mem_overlap_check)
 
+    @expectedFailureMeta  # RuntimeError not raised
```
Comment (kshitij12345) [inline test/test_torch.py:6456]: Note expectedFailureMeta

File: test/test_torch.py [hunk 13]
```diff
@@ -6461,7 +6464,8 @@ def test_copy_mem_overlap(self, device, dtype):
         self.unary_check_input_output_mem_overlap(
             doubles, sz, lambda input, out: out.copy_(input))
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # RuntimeError not raised
```
Comment (kshitij12345) [inline test/test_torch.py:6467]: Note expectedFailureMeta

File: test/test_torch.py [hunk 14]
```diff
@@ -6491,7 +6495,8 @@ def test_index_copy_mem_overlap(self, device):
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             ind.index_copy_(0, ind.clone(), ind)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # Warning not triggered
```
Comment (kshitij12345) [inline test/test_torch.py:6498]: Note expectedFailureMeta

File: test/test_torch.py [hunk 15]
```diff
@@ -6503,15 +6508,17 @@ def test_index_fill_mem_overlap(self, device):
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             ind.index_fill_(0, ind, 0)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # RuntimeError not raised
```
Comment (kshitij12345) [inline test/test_torch.py:6511]: Note expectedFailureMeta

File: test/test_torch.py [hunk 16]
```diff
@@ -6503,15 +6508,17 @@ def test_index_fill_mem_overlap(self, device):
...
+    @onlyNativeDeviceTypes
     def test_shift_mem_overlap(self, device):
         x = torch.rand(3, device=device)
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             x[:-1] <<= x[1:]
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             x[:-1] >>= x[1:]
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # RuntimeError not raised
```
Comment (kshitij12345) [inline test/test_torch.py:6520]: Note expectedFailureMeta

File: test/test_torch.py [hunk 17]
```diff
@@ -6525,7 +6532,8 @@ def test_bernoulli_mem_overlap(self, device):
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             torch.bernoulli(torch.rand_like(x), out=x)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # RuntimeError not raised
```
Comment (kshitij12345) [inline test/test_torch.py:6535]: Note expectedFailureMeta

File: test/test_torch.py [hunk 18]
```diff
@@ -6544,7 +6552,8 @@ def test_put_mem_overlap(self, device):
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             ind.put_(ind.clone(), ind)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # UserWarning not triggered
```
Comment (kshitij12345) [inline test/test_torch.py:6555]: Note expectedFailureMeta

File: test/test_torch.py [hunk 19]
```diff
@@ -6563,7 +6572,8 @@ def test_index_put_mem_overlap(self, device):
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             ind.index_put_((ind.clone(),), ind)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # UserWarning not triggered
```
Comment (kshitij12345) [inline test/test_torch.py:6575]: Note expectedFailureMeta

File: test/test_torch.py [hunk 20]
```diff
@@ -6589,7 +6599,8 @@ def test_masked_select_mem_overlap(self, device):
         with self.assertRaisesRegex(RuntimeError, 'unsupported operation'):
             torch.masked_select(mask.clone(), mask, out=mask)
 
-    @onlyOnCPUAndCUDA
+    @expectedFailureMeta  # RuntimeError not raised
```
Comment (kshitij12345) [inline test/test_torch.py:6602]: Note expectedFailureMeta

File: torch/testing/_internal/common_device_type.py [hunk 21]
```diff
@@ -906,12 +906,12 @@ def multi_fn(slf, devices, *args, **kwargs):
 
         return multi_fn
 
-# Only runs the test on the CPU and CUDA (the native device types)
-def onlyOnCPUAndCUDA(fn):
+# Only runs the test on the native device type (currently CPU, CUDA, Meta)
```
Comment (kshitij12345) [inline torch/testing/_internal/common_device_type.py:909]: Main update

**Comment by kshitij12345** [comment]
gentle ping @mruberry

File: torch/testing/_internal/common_device_type.py [hunk 22]
```diff
@@ -906,12 +906,14 @@ def multi_fn(slf, devices, *args, **kwargs):
...
+def onlyNativeDeviceTypes(fn):
```
Comment (mruberry) [inline torch/testing/_internal/common_device_type.py:910]: Nice change

File: torch/testing/_internal/common_device_type.py (hunk 21 above)
Comment (mruberry) [reply torch/testing/_internal/common_device_type.py:909]: Thanks for the review pointers ;)

File: torch/testing/_internal/common_device_type.py [hunk 23]
```diff
@@ -187,8 +187,8 @@
 #         In addition to accepting multiple dtypes, the @dtypes decorator
 #         can accept a sequence of tuple pairs of dtypes. The test template
 #         will be called with each tuple for its "dtype" argument.
-#     - @onlyOnCPUAndCUDA
-#         Skips the test if the device is not a CPU or CUDA device
+#     - @onlyNativeDeviceTypes
+#         Skips the test if the device is not native device type (currently CPU, CUDA, Meta)
```
Comment (mruberry) [inline torch/testing/_internal/common_device_type.py:191]: "not native" -> "not a native"

**Review by mruberry — APPROVED** [review]
Thanks for being so thorough and adding the review comments, @kshitij12345!

This looks good to me (and it makes this decorator consistent with onlyCPU and onlyCUDA by removing the "On"). I'll make an internal FB announcement, too.

cc @ezyang -- is this what you were expecting?

fyi @zou3519 this will break functorch, which is importing the onlyOnCPUAndCUDA decorator

The ASAN timeout is unrelated but this will need a rebease.

**Comment by ezyang** [comment]
> cc @ezyang -- is this what you were expecting?

Err, I don't know what I was expecting! Seems... reasonable, I guess?

**Comment by zou3519** [comment]
> fyi @zou3519 this will break functorch, which is importing the onlyOnCPUAndCUDA decorator

Is there a replacement we could use?

**Comment by mruberry** [comment]
> > fyi @zou3519 this will break functorch, which is importing the onlyOnCPUAndCUDA decorator
> 
> Is there a replacement we could use?

The `onlyNativeDeviceTypes` decorator introduced with this PR. 

> > fyi @zou3519 this will break functorch, which is importing the onlyOnCPUAndCUDA decorator
> 
> Is there a replacement we could use?

Yes -- but would you prefer to run meta, cpu, and cuda variants of your tests or maintain just running the cpu and cuda varaints?

**Pull request closed without merging**